Fix for points straddling the international date line

cffk opened the pull request:
This fixes the distance calculation for such points by reducing the
longitude difference to the interval [-180, 180].  It also uses the
library functions sinh and hypot where appropriate.

cffk committed:
Fix for points straddling the international date line

This fixes the distance calculation for such points by reducing the
longitude difference to the interval [-180, 180].  It also uses the
library functions sinh and hypot where appropriate.

danpat commented:
The use of `sinh` is nice, and seems to speed things up.  But `std::hypot` is slower than the exiting use of `std::sqrt`, but almost exactly the same amount :-(  The bounds and overflow checking added by `std::hypot` is of dubious value here because we're dealing with such known small numbers.  On the plus side, the intent is clearer.

```
 before sinh: 10.519254s wall, 10.500000s user + 0.000000s system = 10.500000s CPU (99.8%)
 after sinh: 7.510087s wall, 7.420000s user + 0.030000s system = 7.450000s CPU (99.2%)
 
 before hypot: 7.336395s wall, 7.270000s user + 0.030000s system = 7.300000s CPU (99.5%)
 after hypot: 9.925384s wall, 9.900000s user + 0.000000s system = 9.900000s CPU (99.7%)
```

(these results were under default optimization, ratios seem to stay about the same under `-O3`)

<details>
  <summary>quickbench.cc</summary>

```
#include <boost/timer/timer.hpp>
#include <iostream>
#include <cmath>

double method_A(int n) {
return  std::atan(0.5 * (std::exp(n) - std::exp(-n))) * 180. / M_PI;
}

double method_B(int n) {
return  std::atan(std::sinh(n)) * 180. / M_PI;
}

double method_C(double n) {
return std::sqrt(n*n + (n+1)*(n+1));
}

double method_D(double n) {
return std::hypot(n, n+1);
}


#define LOOPCOUNT 1000000000

int main(void) {

   {
       boost::timer::auto_cpu_timer t;
       double sum = 0;
       for (unsigned i=0; i<LOOPCOUNT; i++) {
         sum += method_A(i);
       }
       std::cout << sum << std::endl;
    }
   {
       boost::timer::auto_cpu_timer t;
       double sum = 0;
       for (unsigned i=0; i<LOOPCOUNT; i++) {
         sum += method_B(i);
       }
       std::cout << sum << std::endl;
    }
   {
       boost::timer::auto_cpu_timer t;
       double sum = 0;
       for (unsigned i=0; i<LOOPCOUNT; i++) {
         sum += method_C(i);
       }
       std::cout << sum << std::endl;
    }
   {
       boost::timer::auto_cpu_timer t;
       double sum = 0;
       for (unsigned i=0; i<LOOPCOUNT; i++) {
         sum += method_D(i);
       }
       std::cout << sum << std::endl;
    }

}
```

</summary>

cffk commented:
`sinh` will also be more accurate for small args (not that this will matter here).  If you do revert the `hypot` change, I recommend you retain it as a comment (noting that it is slower).

mourner reviewed (changes requested):
This looks great!

1. Needs some tests that cross the date line.
2. Are there any other methods beside `distance` that need an equivalent fix for consistency?

cffk committed:
Apply longitude fix to other functions.

Basically, all longitude differences are reduced to [-180,180].
Fix blunder in use of remainder (2nd arg should be 360!).

cffk commented:
I think I have the longitude fix applied everywhere it's needed.  I'm not familiar with this testing framework, so I recommend that someone else add more tests.

Currently the longitude values can be anything -- and that's OK since the angle reduction is done on the fly.  On the other hand, latitude values outside the range [-90,90] can be returned (e.g., by offset); arguably these should be clipped to the allowed range.

I also simplified `bearing`.  `atan2` behaves properly (result is in range [-180,180] and `x=0`, `y=0` is handled).

Oh, and I reverted the `hypot` change (and added a comment).

mourner commented:
@cffk btw, is there a simple alternative to `std::remainder` that we could use in the [reference repo (in JS)](https://github.com/mapbox/cheap-ruler)? We can use its definition (`a - Math.round(a / b) * b`) but I wonder if there was a simpler formula for the purpose.

cffk commented:
I recommend `(a - Math.round(a / b) * b) `.  (Alternatively use `a % b` and then check whether the result isn't in `[-b/2, b/2]`.)  Perhaps it's worth defining a `longitude_difference` function to localize the actual implementation?

cffk committed:
Add tests for distance, area, and bearing for the longitude-wrap case.

cffk committed:
Localize the definition of longitude difference with long_diff.
(Perhaps simplifying the backporting to JS.)

cffk commented:
Commentary:

An alternative fix to the longitude wrapping issue would be to enforce
the cylindrical topology in the point class (e.g., normalizing the
longitude to `[-180, 180]` in the constructor).  Then long_diff could be
defined as

    double d = a - b;
    return d < -180 ? d + 360 : (d > 180 ? d - 180: d);

It's not clear whether this is an improvement and once you go down this
path there's the temptation to enforce latitude in `[-90, 90]` and to
over-think the whole problem.

cffk commented:
This is ready to merge, I believe.  Probably more could be done on the tests, but these are going to be revamped as part of #13.  By the way, what's the provenance of the datasets used for testing?  How were the "correct" results obtained?

mourner commented on include/mapbox/cheap_ruler.hpp:
```diff
@@ -348,6 +340,9 @@ class CheapRuler {
     double kx;
+    static double long_diff(double a, double b) {
+        return std::remainder(a - b, 360);
+    }
```
Nitpick: can we rename this to `longDiff` for consistency with other method names?

cffk replied:
Done

cffk committed:
long_diff renamed to longDiff.

vng committed:
Merge branch 'master' into longitude-fix
